### PR TITLE
feat: reworked closing mechanism for ExternalJarClassLoader

### PR DESCRIPTION
**Pull request by igpetrov**
## Description

feat: reworked closing mechanism for ExternalJarClassLoader

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

Related to: https://github.com/camunda/camunda/issues/21559

### CODE REVIEW

File: zeebe/util/src/test/java/io/camunda/zeebe/util/jar/ExternalJarClassLoaderTest.java
Comment (npepinpe) [inline zeebe/util/src/test/java/io/camunda/zeebe/util/jar/ExternalJarClassLoaderTest.java:125]: :x: Please use AssertJ for assertions like in the rest of Zeebe code :+1: You can use `assertThatCode(() -> ...)` to test exceptions.

File: zeebe/util/src/test/java/io/camunda/zeebe/util/jar/ExternalJarClassLoaderTest.java
Comment (npepinpe) [inline zeebe/util/src/test/java/io/camunda/zeebe/util/jar/ExternalJarClassLoaderTest.java:1]: :x: All three variants of `shouldCloseWhenDefaultCloseInvoked` can be a parameterized test, e.g.

```java
@ParameterizedTest
@NullSource // used to simulate the default case
@ValueSource(booleans = {true, false})
void shouldCloseWhenInvoked(final Boolean param) {
  // given
  final var serviceClass = ExternalService.createUnloadedExporterClass();
  final var jarFile = serviceClass.toJar(new File(tempDir, "service.jar"));
  final var classLoader = ExternalJarClassLoader.ofPath(jarFile.toPath());

  // when
  if (param == null) {
    classLoader.close();
  } else {
    classLoader.close(param);
  }

  // then
  assertThatCode(() -> classLoader.loadClass(ExternalService.CLASS_NAME)).isInstanceOf(ClassNotFoundException.class);
}
```

I'm also OK with having one test for the default case (i.e. no args), and a parameterized for the true/false one. I think more purist people would prefer this :smile:

File: zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ExternalJarClassLoader.java
Comment (npepinpe) [inline zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ExternalJarClassLoader.java:69]: :rocket:

File: zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ExternalJarClassLoader.java
Comment (npepinpe) [inline zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ExternalJarClassLoader.java:58]: :wrench: I would describe the consequence in clearer terms, e.g. "Closing the class loader may cause future calls to load classes to fail with ClassNotFoundException" or something. I know it's what you wrote, but I'm not sure if someone who has little context would understand it the way you phrased it :shrug: Just a suggestion

**Review by npepinpe — APPROVED** [review]
:+1: Please address all comments before merging, but we don't need a second review.

File: zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ExternalJarClassLoader.java
Comment (igpetrov) [reply zeebe/util/src/main/java/io/camunda/zeebe/util/jar/ExternalJarClassLoader.java:58]: Reworder to `Closing the class loader may cause future class loading calls to fail with ClassNotFoundException`

File: zeebe/util/src/test/java/io/camunda/zeebe/util/jar/ExternalJarClassLoaderTest.java
Comment (igpetrov) [reply zeebe/util/src/test/java/io/camunda/zeebe/util/jar/ExternalJarClassLoaderTest.java:1]: 3 is fine for me :)

**Pull request merged**
